packages/hash/api/src/model/knowledge/link.model.ts

```diff
@@ -111,19 +110,191 @@ export default class {
          *   https://app.asana.com/0/1202805690238892/1202890446280569/f
          */
         ownedById: createdBy,
+        index,
         linkTypeId: linkTypeModel.schema.$id,
         targetEntityId: targetEntityModel.entityId,
       },
     );
 
-    return LinkModel.fromPersistedLink(graphApi, {
+    /** @todo: this should be returned directly from the `createLink` method */
+    const persistedLink = {
       inner: link,
       ownedById: createdBy,
+    };
+
+    return LinkModel.fromPersistedLink(graphApi, persistedLink);
+  }
+
+  /**
+   * Create a link between a source and a target entity using a specific link
+   * type.
+   *
+   * @param params.accountId the accountId of the account creating the link
+   * @param params.sourceEntityModel the source entity of the link
+   * @param params.linkTypeModel the Link Type of the link
+   * @param params.targetEntityModel the target entity of the link
+   */
+  static async create(
+    graphApi: GraphApi,
+    params: LinkModelCreateParams,
+  ): Promise<LinkModel> {
+    const { sourceEntityModel, linkTypeModel, createdBy } = params;
+    const siblingLinks = await sourceEntityModel.getOutgoingLinks(graphApi, {
+      linkTypeModel,
+    });
+
+    /** @todo: rely on Graph API validation instead of manually checking whether sibling links are ordered */
+    const hasOrderedSiblingLink = siblingLinks[0]?.index !== undefined;
+
+    const index = hasOrderedSiblingLink
+      ? // if siblings are ordered and an index is provided, use the provided index
+        params.index ??
+        // if siblings are ordered and no index is provided, default to the end of the list of links
+        siblingLinks.length
+      : siblingLinks.length === 0
+      ? // if siblings are not ordered because there are no siblings, allow the link to be ordered
```

Comment:
I'm not sure what this means. Surely the "allowing" should be checked against the type rather than the siblings? I appreciate the Graph API should tell us if we're doing something wrong but in the meantime I don't think we want to be allowing an ordered link if the type says it's not ordered, right?

Reply:
Is that what:
> Cannot make an un-ordered link ordered

is supposed to be about inside `updateWithoutUpdatingSiblings`?

Reply:
> in the meantime I don't think we want to be allowing an ordered link if the type says it's not ordered, right?

If you're suggesting that the workspace backend should implement schema validation like this before the Graph API, only to be replaced by the upcoming implementation in the Graph API, then I disagree. I think it is duplicating work.

The implementation made in this PR is temporary, which is why some in-perfect assumptions like this one have made it into the PR. Would you like me to add a comment to clarify that ideally the schema of the source entity should be checked to see if the link is ordered?

Or do you still feel strongly that we should implement this now (even if we are just going to re-implement it in the Graph API)? In which case I can do it, but it would be inconsistent with how we've implemented other things in the workspace backend (where we are waiting for the Graph API to correctly validate things).

Reply:
Hm. I guess I'm unclear about how much of the future logic is going to live in the workspace BE vs the graph layer. Schema validation and such should definitely happen on the Graph, that's not to say that the workspace _can't_ do validation for faster feedback on failures.

I don't feel strongly about this, especially because, as you've identified, this is temporary code. I just don't want to see us fighting super weird to debug bugs because we're doing things like checking the siblings index to see if something is ordered, rather than just checking the schema and seeing `ordered: true` and trusting that. (Hoping the Graph API will enforce it later on)

Reply:
I'm going to address this in a follow-up: https://app.asana.com/0/1202805690238892/1203045927094838/f

Reply:
I'll leave this conversation open then to make it more discoverable :)